File: Week_01/java/src/_155_MinStack.java
import java.util.LinkedList;

/**
 * 第一遍：2020/08/13周四 ✅
 * 第二遍：2020/08/14周五 ✅
 * 第三遍：2020/08/16周日 ✅
 * 第四遍：2020/07/28周二
 * 第五遍：2020/08/09周日
 * 第六遍：2020/08/09周一
 * 第七遍：2020/08/12周三
 */
class _155_MinStack {
    class MinStack {
        LinkedList<Integer> stack = new LinkedList<>();
        LinkedList<Integer> minStack = new LinkedList<>();

        /**
         * initialize your data structure here.
         */
        public MinStack() {

        }

        public void push(int x) {
            stack.push(x);
            if (!minStack.isEmpty()) {
                x = Math.min(x, minStack.peek());
            }
            minStack.push(x);
        }

        public void pop() {
            if (!stack.isEmpty()) {
                stack.pop();
                minStack.pop();
            }
        }

        public int top() {
            if (!stack.isEmpty()) {
                return stack.peek();
            }
            return -1;
        }

        public int getMin() {
            if (!minStack.isEmpty()) {
                return minStack.peek();
            }
            return -1;
        }
    }
}

File: Week_01/kotlin/src/_155_MinStack.kt
import java.util.*

class MinStack() {
    /** initialize your data structure here. */
    val stack = LinkedList<Int>()
    val minStack = LinkedList<Int>()

    fun push(x: Int) {
        stack.push(x)
        minStack.push(if (minStack.isEmpty() || minStack.peek() > x) x else minStack.peek())
    }

    fun pop() {
        stack.pop()
        minStack.pop()
    }

    fun top(): Int {
        return stack.peek()
    }

    fun getMin(): Int {
        return minStack.peek()
    }

}
